### PR TITLE
Add support to sort compilation errors

### LINKED ISSUE #14628: Compile errors not ordered by line number
Consider the following program
```ballerina
import ballerina/io; 

type A object {
    B b;
    string a;
};

type B object {
    A a;
    string b;
};

public function main() {
    A?[] ar = [];
    ar[10] = {a: "Foo"};
    io:println(ar);

    B[] br = [];
    br[10] = {b: "Bar"};
    io:println(br);

    T1[] tr = [];
    tr[10] = "A";
    io:println(tr);
}
```
The errors given for the above:
```
error: .::14620.bal:15:14: invalid literal for type 'A?'
error: .::14620.bal:19:14: invalid usage of record literal with type 'B'
error: .::14620.bal:22:5: unknown type 'T1'
error: .::14620.bal:22:15: array element type 'other' does not have an implicit initial value, use 'other?'
error: .::14620.bal:4:5: uninitialized field 'b'
error: .::14620.bal:5:5: uninitialized field 'a'
error: .::14620.bal:9:5: uninitialized field 'a'
error: .::14620.bal:10:5: uninitialized field 'b'
```
The above should be ordered by line number.

### PR DESCRIPTION
**Pull request by Shan1024**
## Purpose
This PR adds support to sort compilation errors.

The sorting is done in the following order.

1. Package name
2. Package version
3. Compilation unit name
4. Start line
5. Start column
6. Error message

Fixes #14628

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Required Balo version update
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/diagnotic/BDiagnostic.java [hunk 1]
```diff
@@ -79,10 +79,33 @@ public String getMessage() {
...
-    
+
+    @Override
+    public int hashCode() {
+        return kind.hashCode() + pos.hashCode() + msg.hashCode() + code.hashCode();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (!(obj instanceof BDiagnostic)) {
```
Comment (hasithaa) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/diagnotic/BDiagnostic.java]: One of the best practice is to check reference equality first. Shall we add that?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/diagnotic/BDiagnosticSource.java [hunk 2]
```diff
@@ -49,4 +49,26 @@ public String getPackageVersion() {
     public String getCompilationUnitName() {
         return cUnitName;
     }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (!(obj instanceof BDiagnosticSource)) {
```
Comment (hasithaa) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/diagnotic/BDiagnosticSource.java]: Same validation here as well. https://github.com/ballerina-platform/ballerina-lang/pull/18233/files#r318003438

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/diagnotic/BDiagnostic.java (hunk 1 above)
Comment (Shan1024) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/diagnotic/BDiagnostic.java]: Resolved with https://github.com/ballerina-platform/ballerina-lang/pull/18233/commits/eaa8b93163c4386f1dd358a359e7a8ec655d8e4a.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/diagnotic/BDiagnosticSource.java (hunk 2 above)
Comment (Shan1024) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/diagnotic/BDiagnosticSource.java]: Resolved with https://github.com/ballerina-platform/ballerina-lang/pull/18233/commits/eaa8b93163c4386f1dd358a359e7a8ec655d8e4a.

**Pull request merged**